Lewis

lew101 opened the pull request:
WIP:
Dramatically over thought this so I'm going to rework it

lew101 committed:
Merge branch 'master' into Lewis

lew101 committed:
Added the start of the solution to minitask 4

So, it picks a location 3m away really nicely and then starts heading towards there and in theory, it should stop but I haven't got all the bugs out of the 2D distance measuring yet. Object or wall detection haven't been implemented yet.

bhenaghan commented on minitask2/Lewis/walk.py:
```diff
@@ -0,0 +1,226 @@
+FORWARD_SPEED = 0.125
+TURN_SPEED = ((2 * PI) / 360) * 25  # 25 Deg/s
+WALK_DISTANCE = 3.0
+OFFSET_ERROR_LINEAR = 0.01  # Allowable offset from target is 0.01 meters
```
Nice, I really like that these are all constants

bhenaghan commented on minitask2/Lewis/walk.py:
```diff
@@ -0,0 +1,226 @@
+class Walk():
+
+    def __init__(self):
+        rospy.init_node('Walk', anonymous = True)
```
Shouldn't have spaces around named parameters.

bhenaghan commented on minitask2/Lewis/walk.py:
```diff
@@ -0,0 +1,226 @@
+        self.turn_left = Twist()
+        self.turn_left.angular.z = TURN_SPEED
+        self.turn_right = Twist()
+        self.turn_right.angular.z = (-TURN_SPEED)
```
pretty sure the brackets are redundant here

bhenaghan commented on minitask2/Lewis/walk.py:
```diff
@@ -0,0 +1,226 @@
+        self.travel_angle = self.target["delta_theta"]
+
+        rospy.spin()
+
```
normally only one newline between methods within a class

bhenaghan commented on minitask2/Lewis/walk.py:
```diff
@@ -0,0 +1,226 @@
+        Waffle moves forward 3 meters, randomly chooses a direction and rotates to continue the random walk. 
+        This behaviour is overriden by wall and obstacle detection.
+        '''
+
```
There needs to be a `pass` here otherwise your code won't run

bhenaghan commented on minitask2/Lewis/walk.py:
```diff
@@ -0,0 +1,226 @@
+        Helper method to determine if an obstacle is present within 0.5m of the Waffle. 
+        Publishes to the alert topic.
+        '''
+
```
Again, need a `pass`

bhenaghan commented on minitask2/Lewis/walk.py:
```diff
@@ -0,0 +1,226 @@
+        self.vel_publisher.publish(Twist())
+
+
+    def is_obstacle(self): # MOVE TO ANOTHER NODE
```
Okay this one is **super** anal but you need two spaces before the `#` then one after. You only have one before.

bhenaghan commented on minitask2/Lewis/walk.py:
```diff
@@ -0,0 +1,226 @@
+        Helper method used to detect if a wall is present on the right side. 
+        Publishes to the alert topic.
+        '''
+
```
`pass`

bhenaghan commented on minitask2/Lewis/walk.py:
```diff
@@ -0,0 +1,226 @@
+
+        self.plot_trajectory(odom)
+
+        self.drive()
```
I think calling `drive` would be better in a laserscan callback, rather than odometry (as you need to make urgent decisions based off of the scan data, but odometry data isn't really very relevant most of the time)

bhenaghan commented on minitask2/Lewis/walk.py:
```diff
@@ -0,0 +1,226 @@
+                if self.curr_pose["theta"] > abs(self.travel_angle):
+                    self.turning = False
+                    self.drive_stop()
+                    print("[LOG]    Current angle: " + str(self.curr_pose["theta"]))
```
doing this with `"abc {}".format(def)` or `"abc {0}" % (def)` is more pythonic/readable than doing a normal string concat imo

bhenaghan commented on minitask2/Lewis/walk.py:
```diff
@@ -0,0 +1,226 @@
+                    else:
+                        self.drive_turn_right()
+            else:
+                distance_remaining = self.get_distance_remaining(self.curr_pose["x"], self.curr_pose["y"], self.target["x"], self.target["y"])
```
put these params on separate lines to make this line a lot shorter.

bhenaghan commented on minitask2/Lewis/walk.py:
```diff
@@ -0,0 +1,226 @@
+                    print("[LOG]    Driving complete")
+                    self.drive_stop()
+                    self.turning = True
+                    self.target = self.get_new_target
```
You need brackets after this to call the method, you've just set `self.target` to be a reference to the method `get_new_target` - you never actually call the method.

bhenaghan commented on minitask2/Lewis/walk.py:
```diff
@@ -0,0 +1,226 @@
+
+
+    def get_new_target(self, travel_distance = WALK_DISTANCE):
+        # set target position using polar coordinate maths
```
This is all very complicated! pick a theta, turn to it and then drive forwards for 3m!

bhenaghan commented on minitask2/Lewis/walk.py:
```diff
@@ -0,0 +1,226 @@
+            
+        elif (x > 0 and y < 0):
+            return angle
+        
```
What if `y = 100`, `x=0`, shouldn't the angle be about 90 degrees. Might be wrong here - haven't really thought it through, but my intuition is that there are cases where either x or y can be 0 and the angle should not be 0.

lew101 replied:
Honestly, I also hadn't thought about this. I'd have to plot arctan and have a look at some point. For now though, this shouldn't be required anymore so I'm going to leave it. A good spot though!

bhenaghan commented on minitask2/Lewis/walk.py:
```diff
@@ -0,0 +1,226 @@
+            if current_y > target_y:
+                return sqrt((current_x - target_x) ** 2 + (current_y - target_y) ** 2)
+            else:
+                return sqrt((current_x - target_x) ** 2 + (target_y - current_y) ** 2)
```
You could remove all of the if statements by just using `abs`!

bhenaghan commented on minitask2/Lewis/walk.py:
```diff
@@ -0,0 +1,226 @@
+    
+    def get_distance_remaining_1d(self, current_x, target_x):
+        if current_x > target_x:
+            return sqrt((current_x - target_x) ** 2)
```
Again, use `abs`! You're also squaring something then immediately `sqrt`ing it, you could replace this entire method with `return abs(current_x - target_x)`!

bhenaghan commented on minitask2/Lewis/walk.py:
```diff
@@ -0,0 +1,226 @@
+    try:
+        Walk()
+    except rospy.ROSInterruptException:
+        pass
```
normally a good idea to have a newline at the end of the file

BenH14 reviewed (changes requested):
I see what you mean about over-complicating it! Start with the easy stuff (eg obstacle avoidance) and work up to the random walk and wall following. Remember that you're meant to be able to finish this entire task in a couple of hours!

Also, my comments are super nit-picky (One of them is particularly ridiculous, but it is something mentioned in `pep8`).

lew101 committed:
Merge branch 'master' into Lewis

lew101 committed:
Completed the code for random walk and tidied

finally

lew101 committed:
Obstacle avoidance has been implemented in a second node

The obstacle avoidance node runs at the same time as the random walk one and overrides its behaviour when there is an object near

lew101 committed:
Wall following implemented but is behaving poorly with object detection

I can't figure out which one should take priority (although probably object detection)

lew101 reviewed (commented):
All changed and now nearly working in a fun and interesting way

psymaco commented on minitask2/Lewis/obstacle_avoidance.py:
```diff
@@ -0,0 +1,143 @@
+        self.forward.linear.x = FORWARD_SPEED
+       
+        self.turn_left = Twist()
+        self.turn_left.linear.x = 0
```
perhaps you should define the zero integers at the beginning?

bhenaghan commented on minitask2/Lewis/obstacle_avoidance.py:
```diff
@@ -0,0 +1,143 @@
+        '''
+        min_val_angle = 0
+        min_val = 50
+        for i in range(300, 359):
```
you could do this in one for loop by doing `range(-60,60)` when you index with a negative value in python it counts from the end of the list.

bhenaghan commented on minitask2/Lewis/obstacle_avoidance.py:
```diff
@@ -0,0 +1,143 @@
+
+        if not self.blocked:
+            if min_val < MIN_DISTANCE:
+                if self.blocking == False:
```
You can simplify this to just be `if not self.blocking:` like line `76`

bhenaghan commented on minitask2/Lewis/obstacle_avoidance.py:
```diff
@@ -0,0 +1,143 @@
+        if not self.direction_set:
+            rospy.loginfo("Obstacle avoidance")
+            if min_val_angle < 90:
+                self.direction = 2
```
I think this `2` really needs to be a constant to make this clear.

bhenaghan commented on minitask2/Lewis/walk.py:
```diff
@@ -0,0 +1,198 @@
+        self.travel_angle = 0.0
+        
+        # For odometry and position tracking
+        self.curr_pose = {}
```
All the initialisations can be done in the dict definition: `self.curr_pose = {"theta": 0.0, "x": 0.0, "y": 0.0}`

bhenaghan commented on minitask2/Lewis/walk.py:
```diff
@@ -0,0 +1,198 @@
+        while not self.end:
+            if not self.blocked:
+                if self.turning:
+                    if (self.travel_angle < 0 and self.curr_pose["theta"] > self.travel_angle) or (self.travel_angle >= 0 and self.curr_pose["theta"] < self.travel_angle):
```
This line is a monster, we could defo simplify it and/or split into separate lines.

bhenaghan commented on minitask2/Lewis/walk.py:
```diff
@@ -0,0 +1,198 @@
+                        self.turning = False
+                        self.drive_stop()
+                else:
+                    progress = sqrt((self.last_pose["x"] - self.curr_pose["x"])** 2 + (self.last_pose["y"] - self.curr_pose["y"])** 2)
```
**nitpick:** you should have a space before the `**`

bhenaghan commented on minitask2/Lewis/obstacle_avoidance.py:
```diff
@@ -0,0 +1,143 @@
+        Sets the blocking flag if higher priority actions are taking place
+        '''
+        rospy.loginfo("Blocking message received")
+        if str(data.data) == str(3):
```
I don't know why you're casting to string here? the `str(3)` is defo redundant as you could just do `"3"`.

bhenaghan commented on minitask2/Lewis/wall_follow_right.py:
```diff
@@ -0,0 +1,132 @@
+        '''
+        min_val_right = 50
+        for i in range(70, 110):
+            if scan.ranges[i] < min_val_right:
```
you can replace these two lines with `min_val_right = min(min_val_right, scan.ranges[i]` - Its not absolutely better but most people prefer it. It also has the advantage of branchless programming, which leads to an almost imperceptible increase in performance! branchless programming is generally a good habit to get into though, especially in python examples like this, as min() can be implemented as a fast bit of `c` code in some interpreters as its part of the standard library.

bhenaghan commented on minitask2/Lewis/wall_follow_right.py:
```diff
@@ -0,0 +1,132 @@
+                    self.drive_stop()
+                    self.blocking = True
+                    self.direction_set = False
+                    self.blocking_publisher.publish(1)
```
lines `86, 87, 96, 97` really mean nothing when its just magic numbers (not constants).

BenH14 reviewed (commented):
This code is a great improvement, and I would say I'm happy with it overall. I do however have a few qualms about the broad design.

I think your intentions are good with splitting the code up into multiple nodes, this is something that is often done with real robotics projects when the behaviours are decently complex. I do think that if you're going to do it this way you really need a 4th node to handle state and when each of the nodes should be giving velocity commands. The ideal way for this to work would be that  each behaviour node publishes a Twist commands to a topic that the controller (4th node) subscribes to, it could then make decisions about which velocity command to publish to `/vel_cmd`. This should make each node much less stateful - something that I would say is the main problem with your current code. 

With ROS being a microservice architecture you can't really rely on your nodes being executed fairly or in any meaningful order,  so the current system of blocking could lead to certain nodes never running (another node always gets there first and locks before it has a chance to block and run). In a bigger robotics project you'll often have something like a controller which makes decisions about which behaviour to run (essentially running a state machine), which will then make requests to behaviour nodes for their required paths. The paths will then be handed to a path following node, which figures out the right velocity commands to send. 

The difference between this kind of architecture and what you've implemented is that there's a clear hierarchy of nodes. In your implementation each node seems to scramble to lock without any authority handling when this should be allowed.
